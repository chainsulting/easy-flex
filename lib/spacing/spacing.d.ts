import { FC, HTMLAttributes } from 'react';
import { IDistance, IViewportThreshold } from '../types';
export interface ISpacingProps extends HTMLAttributes<HTMLDivElement> {
    /** If true, height and width are flipped if flipping has been enabled. */
    flip?: boolean;
    /** Enables flipping. */
    flipEnabled?: boolean;
    /** Sets the viewport threshold. The content will be flipped if the viewport's width is smaller than the threshold. If no threshold is set, the fallback threshold is used. */
    viewportThreshold?: IViewportThreshold;
    /** Component's height. */
    height?: IDistance | number;
    /** Component's width. */
    width?: IDistance | number;
}
export declare const Spacing: FC<ISpacingProps>;
