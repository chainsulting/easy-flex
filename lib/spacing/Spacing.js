"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Spacing = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var react_1 = require("react");
var styled_components_1 = __importDefault(require("styled-components"));
var utils_1 = require("../utils/utils");
var StyledSpacing = styled_components_1.default.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n\tdisplay: flex;\n\tbox-sizing: border-box;\n\tbackground-color: transparent;\n\tmargin: 0;\n\tpadding: 0;\n\tmin-height: ", ";\n\theight: ", ";\n\tmax-height: ", ";\n\tmin-width: ", ";\n\twidth: ", ";\n\tmax-width: ", ";\n"], ["\n\tdisplay: flex;\n\tbox-sizing: border-box;\n\tbackground-color: transparent;\n\tmargin: 0;\n\tpadding: 0;\n\tmin-height: ", ";\n\theight: ", ";\n\tmax-height: ", ";\n\tmin-width: ", ";\n\twidth: ", ";\n\tmax-width: ", ";\n"])), function (_a) {
    var height = _a["data-height"];
    return height;
}, function (_a) {
    var height = _a["data-height"];
    return height;
}, function (_a) {
    var height = _a["data-height"];
    return height;
}, function (_a) {
    var width = _a["data-width"];
    return width;
}, function (_a) {
    var width = _a["data-width"];
    return width;
}, function (_a) {
    var width = _a["data-width"];
    return width;
});
exports.Spacing = (0, react_1.forwardRef)(function (_a, ref) {
    var flip = _a.flip, _b = _a.flipEnabled, flipEnabled = _b === void 0 ? false : _b, viewportThreshold = _a.viewportThreshold, _c = _a.height, height = _c === void 0 ? '0px' : _c, _d = _a.width, width = _d === void 0 ? '0px' : _d, props = __rest(_a, ["flip", "flipEnabled", "viewportThreshold", "height", "width"]);
    var theme = (0, utils_1.useEasyFlexTheme)();
    var displayWidth = (0, utils_1.useDimension)().width;
    var processedHeight = (0, react_1.useMemo)(function () {
        return flipEnabled &&
            (flip ||
                (flip === undefined &&
                    (viewportThreshold !== undefined
                        ? displayWidth < (0, utils_1.getViewportThreshold)(theme, viewportThreshold)
                        : displayWidth < theme.viewport.fallbackThreshold)))
            ? (0, utils_1.getDistance)(theme, width)
            : (0, utils_1.getDistance)(theme, height);
    }, [displayWidth, flip, flipEnabled, height, theme, viewportThreshold, width]);
    var processedWidth = (0, react_1.useMemo)(function () {
        return flipEnabled &&
            (flip ||
                (flip === undefined &&
                    (viewportThreshold !== undefined
                        ? displayWidth < (0, utils_1.getViewportThreshold)(theme, viewportThreshold)
                        : displayWidth < theme.viewport.fallbackThreshold)))
            ? (0, utils_1.getDistance)(theme, height)
            : (0, utils_1.getDistance)(theme, width);
    }, [displayWidth, flip, flipEnabled, height, theme, viewportThreshold, width]);
    return (0, jsx_runtime_1.jsx)(StyledSpacing, __assign({ "data-height": processedHeight, "data-width": processedWidth, ref: ref }, props));
});
exports.Spacing.displayName = 'Spacing';
var templateObject_1;
