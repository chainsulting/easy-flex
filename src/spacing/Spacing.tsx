import React, { forwardRef, HTMLAttributes, memo, useMemo } from 'react';
import styled from 'styled-components';
import { AbsoluteSize, CustomName, Distance, Falsifiable } from '../types';
import { defalsify, getDistance, useEasyFlexTheme } from '../utils/base';
import { FlexItemProps, flexItemStyle, FlexItemStyleProps, useFlexItemStyleProps } from '../utils/flexItem';
import { MiscProps, miscStyle, MiscStyleProps, useMiscStyleProps } from '../utils/misc';

const StyledDiv = styled.div<
	{ 'data-height': AbsoluteSize; 'data-width': AbsoluteSize } & FlexItemStyleProps & MiscStyleProps
>`
	background-color: transparent;
	box-sizing: border-box;
	margin: 0;
	padding: 0;
	min-height: ${({ 'data-height': height }) => height};
	height: ${({ 'data-height': height }) => height};
	max-height: ${({ 'data-height': height }) => height};
	min-width: ${({ 'data-width': width }) => width};
	width: ${({ 'data-width': width }) => width};
	max-width: ${({ 'data-width': width }) => width};
	${flexItemStyle}
	${miscStyle}
`;

export interface SpacingProps<CustomDistance extends CustomName>
	extends HTMLAttributes<HTMLDivElement>,
		FlexItemProps,
		MiscProps<CustomDistance> {
	flip?: boolean;
	/** Component's height. */
	height?: Falsifiable<Distance<CustomDistance>>;
	/** Component's width. */
	width?: Falsifiable<Distance<CustomDistance>>;
}

export const createSpacing = <CustomDistance extends CustomName>() => {
	const Spacing = memo(
		forwardRef<HTMLDivElement, SpacingProps<CustomDistance>>(
			(
				{
					alignSelf,
					basis,
					bottom,
					display,
					flex,
					flip,
					grow,
					height,
					left,
					position,
					right,
					shrink,
					top,
					userSelect,
					visibility,
					width,
					...props
				},
				ref
			) => {
				const theme = useEasyFlexTheme();

				const processedHeight = useMemo<AbsoluteSize>(
					() =>
						flip === undefined || flip === false
							? getDistance(theme, defalsify(height) ?? '0px')
							: getDistance(theme, defalsify(width) ?? '0px'),
					[flip, height, theme, width]
				);

				const flexItemStyleProps = useFlexItemStyleProps({ alignSelf, basis, flex, grow, shrink });

				const miscStyleProps = useMiscStyleProps({
					bottom,
					display: defalsify(display) ?? 'flex',
					left,
					position,
					right,
					top,
					userSelect,
					visibility,
				});

				const processedWidth = useMemo<AbsoluteSize>(
					() =>
						flip === undefined || flip === false
							? getDistance(theme, defalsify(width) ?? '0px')
							: getDistance(theme, defalsify(height) ?? '0px'),
					[flip, height, theme, width]
				);

				return (
					<StyledDiv
						data-height={processedHeight}
						data-width={processedWidth}
						{...flexItemStyleProps}
						{...miscStyleProps}
						ref={ref}
						{...props}
					/>
				);
			}
		)
	);
	Spacing.displayName = 'Spacing';
	return Spacing;
};
