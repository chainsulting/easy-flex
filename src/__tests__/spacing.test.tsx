import { render } from '@testing-library/react';
import 'jest-styled-components';
import React from 'react';
import { Spacing } from '../spacing/Spacing';

describe('<Spacing />', () => {
	test('should have the basic css settings', async () => {
		const result = render(<Spacing />);
		expect(result.baseElement).toMatchSnapshot();
	});
});
