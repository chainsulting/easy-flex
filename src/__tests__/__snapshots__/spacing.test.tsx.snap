// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Spacing /> should have the basic css settings 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: transparent;
  margin: 0;
  padding: 0;
  min-height: 0px;
  height: 0px;
  max-height: 0px;
  min-width: 0px;
  width: 0px;
  max-width: 0px;
}

<body>
  <div>
    <div
      class="c0"
      data-height="0px"
      data-width="0px"
    />
  </div>
</body>
`;
